### DISCUSSION TITLE
cli arg parser from lua

### DISCUSSION THREAD

**Discussion by cscherrNT** [mfussenegger/nvim-dap#1160] [Show and tell]
**Tl;Dr** With the provided lua code, you can get a seperate arguments field and pass those arguments to the debugged application!

I was debugging a program that required large cli args, so I made a few functions and a lexer for arguments in lua, maybe someone can need it.

My config for the plugin directly:
```lua
            dap.configurations.cpp = {
                {
                    name = "Launch file",
                    type = "codelldb",
                    request = "launch",
                    program = function()
                        return vim.fn.input("Path to executable: ", vim.fn.getcwd() .. "/", "file")
                    end,
                    args = function()
                        local t = require('custom.utils')
                        return t.tokenize_args(vim.fn.input("args: "))
                    end,
                    cwd = "${workspaceFolder}",
                    -- FIXME: perhaps we can put the stdio somewhere more practical
                    stopOnEntry = false,
                },
            }
```

`require('custom.utils').tokenize_args` is a custom function:
```lua
function(raw)
    -- NOTE: string.gmatch is does not use regex, but a smaller pattern matcher!
    -- A complete regex parser would be larger than lua itself. See
    -- [Programming in Lua 20.2](https://www.lua.org/pil/20.2.html).
    --
    -- Notable differences:
    -- '-' is ungreedy wildcard
    -- '*?' does not work
    -- '|' is not or
    --
    -- This means we're better of implementing the lexer with an algorithm.
    local t = {}
    local current = ""
    local in_str = false
    local str_seek
    for c in string.gmatch(raw, ".") do -- iterate through all chars
        if c == ' ' and not in_str then
            if string.len(current) > 0 then
                table.insert(t, current)
                current = ""
            end
        elseif c == '"' and not in_str then
            in_str = true
            str_seek = '"'
        elseif c == "'" and not in_str then
            in_str = true
            str_seek = "'"
        elseif c == str_seek and in_str then
            in_str = false
            table.insert(t, current)
            current = ""
        else
            current = current .. c
        end
    end
    if string.len(current) > 0 then
        table.insert(t, current)
    end
    return t
end
```

For example, parsing of `"--foo -asfj --data '{\"FOO\": \"BAR\"}' --file=/etc/passwd"` would give
`{ [1] = '--foo', [2] = '-asfj', [3] = '--data', [4] = '{"FOO": "BAR"}', [5] = '--file=/etc/passwd' }`

See [here](https://git.cscherr.de/PlexSheep/lua-basic/src/branch/master/tokenize_args.lua) for full example.

**Comment by cscherrNT** [comment]
I'm not looking for answers or anything, just putting this here in case someone needs something like this.
